File: client-sdk/typescript/test/harness/nats-server.ts
// Local `nats-server` process lifecycle for integration tests.
//
// The binary must be on PATH. We skip tests gracefully when it is not,
// rather than failing the run — that pattern is friendlier for contributors
// who haven't installed `nats-server` yet. Matches the Python SDK's approach.
//
// `start()` composes three things on the command line: the `-a/-p` listen
// flags (always), an optional `-c <config>` — the repo-level identity
// fixtures under IDENTITY_FIXTURES_DIR are port-less on purpose, `-a/-p`
// supply the address, so nothing is templated — and an optional
// `-js -sd <tmpdir>` for JetStream. stderr is captured so a config the
// server rejects fails `start()` fast with the reason, instead of a bare
// 5 s timeout.
//
// Byte-identical to agent-sdk/typescript/test/harness/nats-server.ts — keep
// the two in sync.

import { type ChildProcess, spawn } from "node:child_process";
import { access, constants as fsc, mkdtemp, rm } from "node:fs/promises";
import { connect as tcpConnect, createServer } from "node:net";
import { tmpdir } from "node:os";
import { join as joinPath } from "node:path";
import { fileURLToPath } from "node:url";

/**
 * Repo-level identity fixtures (`test-fixtures/identity/`): throwaway nkeys
 * (`keys.json`), agent-ID parse fixtures, and port-less nats-server configs.
 *
 * Resolved from this file's URL, never from `cwd` — CI runs vitest with
 * cwd = the package dir. Four `..`: `harness/` → `test/` → `<package>/` →
 * `client-sdk/` (or `agent-sdk/`) → repo root.
 */
export const IDENTITY_FIXTURES_DIR = fileURLToPath(
  new URL("../../../../test-fixtures/identity/", import.meta.url),
);

/** Absolute path of a file under {@link IDENTITY_FIXTURES_DIR}. */
export function identityFixture(name: string): string {
  return joinPath(IDENTITY_FIXTURES_DIR, name);
}

export class NatsServerNotAvailableError extends Error {
  constructor(reason: string) {
    super(`nats-server not available: ${reason}`);
    this.name = "NatsServerNotAvailableError";
  }
}

export async function findNatsServerBinary(): Promise<string | null> {
  const pathEnv = process.env["PATH"] ?? "";
  for (const dir of pathEnv.split(":")) {
    if (!dir) continue;
    const candidate = joinPath(dir, "nats-server");
    try {
      await access(candidate, fsc.X_OK);
      return candidate;
    } catch {
      /* try next */
    }
  }
  return null;
}

/** A free TCP port on 127.0.0.1 (bind-to-0, read back, release). */
export async function findFreePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const srv = createServer();
    srv.once("error", reject);
    srv.listen(0, "127.0.0.1", () => {
      const addr = srv.address();
      if (addr && typeof addr === "object") {
        const { port } = addr;
        srv.close(() => resolve(port));
      } else {
        srv.close();
        reject(new Error("unexpected socket address"));
      }
    });
  });
}

function tryTcp(host: string, port: number): Promise<boolean> {
  return new Promise((resolve) => {
    const s = tcpConnect({ host, port });
    s.once("connect", () => {
      s.end();
      resolve(true);
    });
    s.once("error", () => resolve(false));
  });
}

/**
 * Poll until `host:port` accepts a TCP connection. `abortReason`, when
 * given, is consulted on every iteration; a non-null string aborts the wait
 * with that message — used to fail fast once the server process has exited.
 */
export async function waitForTcp(
  host: string,
  port: number,
  timeoutMs: number,
  abortReason?: () => string | null,
): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const reason = abortReason?.() ?? null;
    if (reason !== null) throw new Error(reason);
    if (await tryTcp(host, port)) return;
    await new Promise((r) => setTimeout(r, 50));
  }
  throw new Error(`nats-server did not open port ${port} within ${timeoutMs}ms`);
}

export interface NatsServerStartOptions {
  /**
   * Config file for `-c`. Fixture configs are port-less; the harness's
   * `-a/-p` flags still supply the listen address (verified on 2.12.7).
   */
  readonly configPath?: string;
  /** Enable JetStream (`-js`) on a throwaway store dir (`-sd`), removed by `stop()`. */
  readonly jetstream?: boolean;
}

/** How much of the server's stderr to keep for error messages. */
const STDERR_TAIL_BYTES = 4_096;
const START_TIMEOUT_MS = 5_000;

export class NatsServerProcess {
  #proc: ChildProcess | null = null;
  #port = 0;
  #ready = false;
  #storeDir: string | null = null;
  #stderrTail = "";

  get url(): string {
    if (this.#port === 0) throw new Error("NatsServerProcess.url: not started");
    return `nats://127.0.0.1:${this.#port}`;
  }

  get port(): number {
    return this.#port;
  }

  /** Tail of the server's stderr — config errors land here before `-l` logging starts. */
  get stderrTail(): string {
    return this.#stderrTail;
  }

  /** JetStream store dir when started with `{ jetstream: true }`, else `null`. */
  get storeDir(): string | null {
    return this.#storeDir;
  }

  async start(opts: NatsServerStartOptions = {}): Promise<void> {
    if (this.#proc) return;
    const bin = await findNatsServerBinary();
    if (!bin) {
      throw new NatsServerNotAvailableError("binary not found on PATH");
    }
    this.#port = await findFreePort();
    this.#stderrTail = "";
    this.#ready = false;
    const args = ["-a", "127.0.0.1", "-p", String(this.#port)];
    if (opts.configPath !== undefined) args.push("-c", opts.configPath);
    if (opts.jetstream) {
      this.#storeDir = await mkdtemp(joinPath(tmpdir(), "synadia-nats-js-"));
      args.push("-js", "-sd", this.#storeDir);
    }
    const proc = spawn(bin, args, { stdio: ["ignore", "ignore", "pipe"], detached: false });
    this.#proc = proc;
    proc.stderr?.setEncoding("utf8");
    proc.stderr?.on("data", (chunk: string) => {
      this.#stderrTail = (this.#stderrTail + chunk).slice(-STDERR_TAIL_BYTES);
    });
    // `close` fires after every stdio stream has drained — awaiting it on the
    // failure path guarantees the stderr tail in the error is complete.
    const closed = new Promise<void>((resolve) => proc.once("close", () => resolve()));
    proc.once("exit", (code, signal) => {
      if (this.#ready && this.#proc === proc) {
        // unexpected exit after a successful start
        console.error(
          `nats-server exited unexpectedly (code=${code}, signal=${signal})${this.#tailSuffix()}`,
        );
      }
    });
    const exited = (): string | null =>
      proc.exitCode !== null || proc.signalCode !== null
        ? `nats-server exited before listening (code=${proc.exitCode}, signal=${proc.signalCode})`
        : null;
    try {
      await waitForTcp("127.0.0.1", this.#port, START_TIMEOUT_MS, exited);
      this.#ready = true;
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      await this.stop();
      await Promise.race([closed, new Promise((r) => setTimeout(r, 500))]);
      throw new Error(`${message}; args: ${args.join(" ")}${this.#tailSuffix()}`);
    }
  }

  async stop(): Promise<void> {
    const proc = this.#proc;
    this.#proc = null;
    this.#ready = false;
    if (proc && proc.exitCode === null && proc.signalCode === null) {
      proc.kill("SIGTERM");
      await new Promise<void>((resolve) => {
        const killTimer = setTimeout(() => {
          proc.kill("SIGKILL");
          resolve();
        }, 3_000);
        proc.once("exit", () => {
          clearTimeout(killTimer);
          resolve();
        });
      });
    }
    if (this.#storeDir) {
      await rm(this.#storeDir, { recursive: true, force: true });
      this.#storeDir = null;
    }
  }

  #tailSuffix(): string {
    const tail = this.#stderrTail.trim();
    return tail ? `\n--- nats-server stderr (tail) ---\n${tail}` : "";
  }
}

File: client-sdk/python/tests/harness/nats_server.py
"""Spawn a local `nats-server` for integration tests.

Skips integration tests cleanly if `nats-server` is not on PATH. We do not
bundle or auto-download the server — it's a Go binary the user installs via
`brew install nats-server` (macOS) or their distro's instructions.

:func:`start_server` composes three things on the command line: the
``-a/-p`` listen flags (always, plus ``-m`` for monitoring), an optional
``-c <config>`` — the repo-level identity fixtures under
:data:`IDENTITY_FIXTURES_DIR` are port-less on purpose; ``-a/-p`` supply the
address, so nothing is templated — and an optional ``-js -sd <tmpdir>`` for
JetStream. stderr is captured next to the log so a config the server
rejects fails fast with the reason in the exception, instead of a bare 5 s
timeout.

This copy (client-sdk) is canonical; ``agent-sdk/python/tests/harness/
nats_server.py`` is byte-identical — keep the two in sync.
"""

from __future__ import annotations

import shutil
import socket
import subprocess
import tempfile
import time
from contextlib import closing
from dataclasses import dataclass
from pathlib import Path

#: Repo-level identity fixtures (``test-fixtures/identity/``): throwaway
#: nkeys (``keys.json``), agent-ID parse fixtures, and port-less nats-server
#: configs. Resolved from this file, never from the cwd. Counting up from
#: ``tests/harness/nats_server.py``: ``harness`` → ``tests`` → ``<package>``
#: → ``client-sdk`` (or ``agent-sdk``) → repo root, i.e. ``parents[4]``.
IDENTITY_FIXTURES_DIR = Path(__file__).resolve().parents[4] / "test-fixtures" / "identity"

STARTUP_TIMEOUT_S = 5.0
STDERR_TAIL_LINES = 20


def identity_fixture(name: str) -> Path:
    """Absolute path of a file under :data:`IDENTITY_FIXTURES_DIR`."""
    return IDENTITY_FIXTURES_DIR / name


def find_nats_server() -> str | None:
    """Absolute path to `nats-server`, or None if not installed."""
    return shutil.which("nats-server")


def _pick_free_port() -> int:
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.bind(("127.0.0.1", 0))
        return int(sock.getsockname()[1])


def _pick_free_ports(n: int) -> tuple[int, ...]:
    """Pick *n* distinct free ports atomically.

    Holds all *n* binds simultaneously before closing any, so the
    kernel cannot hand out the same ephemeral port twice across a
    single multi-port allocation. Two back-to-back :func:`_pick_free_port`
    calls would have a vanishingly rare TOCTOU collision when the
    kernel reuses the same port for both — uncommon, but a harness that
    needs distinct ports shouldn't depend on luck. (The window between
    closing the sockets here and nats-server binding them remains; this
    only guarantees the two ports differ from each other.)
    """
    socks = [socket.socket(socket.AF_INET, socket.SOCK_STREAM) for _ in range(n)]
    try:
        for sock in socks:
            sock.bind(("127.0.0.1", 0))
        return tuple(int(sock.getsockname()[1]) for sock in socks)
    finally:
        for sock in socks:
            sock.close()


@dataclass
class RunningServer:
    """A nats-server running as a child process of the test session.

    Exposes both the client port (``url``) and the HTTP monitoring port
    (``monitoring_url``) so tests that want to verify broker-side state
    (e.g. live subscription count) can hit ``/subsz`` directly rather
    than relying on SDK-internal instrumentation.
    """

    url: str
    process: subprocess.Popen[bytes]
    stdout_log: Path
    port: int
    monitoring_port: int
    monitoring_url: str
    #: The server's stderr — a rejected config lands here before ``-l``
    #: logging starts.
    stderr_log: Path
    #: The ``-c`` config file, if any.
    config_path: Path | None = None
    #: The JetStream store dir when started with ``jetstream=True``; removed
    #: by :meth:`stop`.
    store_dir: Path | None = None

    def stderr_tail(self, lines: int = STDERR_TAIL_LINES) -> str:
        """The last *lines* of the server's stderr (empty if none)."""
        try:
            text = self.stderr_log.read_text(errors="replace")
        except OSError:
            return ""
        return "\n".join(text.splitlines()[-lines:])

    def stop(self) -> None:
        if self.process.poll() is None:
            self.process.terminate()
            try:
                self.process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self.process.kill()
                self.process.wait(timeout=2)
        if self.store_dir is not None:
            shutil.rmtree(self.store_dir, ignore_errors=True)


def start_server(
    log_dir: Path, *, config_path: Path | None = None, jetstream: bool = False
) -> RunningServer:
    """Start a fresh nats-server on a free port, logging verbosely to `log_dir`.

    Returns a handle with ``.url`` (nats://127.0.0.1:<client_port>),
    ``.monitoring_url`` (http://127.0.0.1:<monitoring_port>), and a
    ``.stop()`` method. Raises RuntimeError if the server exits before
    listening (the message carries its stderr — the config error) or if
    either port fails to accept connections within 5 seconds.

    ``config_path`` is passed as ``-c``; the ``-a/-p`` flags override any
    listen address in it, so fixture configs stay port-less. ``jetstream``
    adds ``-js -sd <tmpdir>``; the store dir is removed on ``stop()``.

    The monitoring port is enabled unconditionally: it costs one extra
    listening socket and lets tests assert broker-observed truth (e.g.
    subscription counts via ``/subsz``) instead of having to monkey-
    patch SDK internals.
    """
    binary = find_nats_server()
    if binary is None:
        raise RuntimeError("nats-server not on PATH")

    log_dir.mkdir(parents=True, exist_ok=True)
    port, monitoring_port = _pick_free_ports(2)
    log_file = log_dir / f"nats-server-{port}.log"
    stderr_file = log_dir / f"nats-server-{port}.stderr"

    # -DV = debug+verbose; -a = address; -p = client port; -m = HTTP
    # monitoring port. We write to a file so the test can attach it as
    # evidence on failure without racing the subprocess pipe.
    args = [
        binary,
        "-DV",
        "-a",
        "127.0.0.1",
        "-p",
        str(port),
        "-m",
        str(monitoring_port),
        "-l",
        str(log_file),
    ]
    if config_path is not None:
        args += ["-c", str(config_path)]
    store_dir: Path | None = None
    if jetstream:
        store_dir = Path(tempfile.mkdtemp(prefix="synadia-nats-js-"))
        args += ["-js", "-sd", str(store_dir)]

    # Popen dups the descriptor; closing our handle afterwards is fine.
    with stderr_file.open("wb") as stderr_fh:
        proc = subprocess.Popen(args, stdout=subprocess.DEVNULL, stderr=stderr_fh)

    server = RunningServer(
        url=f"nats://127.0.0.1:{port}",
        process=proc,
        stdout_log=log_file,
        port=port,
        monitoring_port=monitoring_port,
        monitoring_url=f"http://127.0.0.1:{monitoring_port}",
        stderr_log=stderr_file,
        config_path=config_path,
        store_dir=store_dir,
    )
    for what, listen_port in (("client", port), ("monitoring", monitoring_port)):
        if not _wait_for_listen(listen_port, timeout=STARTUP_TIMEOUT_S, process=proc):
            server.stop()
            raise RuntimeError(_startup_failure(what, listen_port, args, server))
    return server


def _startup_failure(what: str, port: int, args: list[str], server: RunningServer) -> str:
    code = server.process.poll()
    if code is not None:
        head = f"nats-server exited before listening (code={code})"
    else:
        head = (
            f"nats-server {what} port :{port} did not accept connections "
            f"within {STARTUP_TIMEOUT_S:g}s"
        )
    message = f"{head}; args: {' '.join(args)}; log: {server.stdout_log}"
    tail = server.stderr_tail()
    if tail:
        message += f"\n--- nats-server stderr (tail) ---\n{tail}"
    return message


def _wait_for_listen(
    port: int, *, timeout: float, process: subprocess.Popen[bytes] | None = None
) -> bool:
    """Poll the port until it accepts a TCP connection or the timeout elapses.

    Returns False early when *process* has already exited — a rejected
    config must not cost the full timeout.
    """
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if process is not None and process.poll() is not None:
            return False
        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
            sock.settimeout(0.25)
            try:
                sock.connect(("127.0.0.1", port))
                return True
            except OSError:
                time.sleep(0.05)
    return False
